### PR TITLE
feat(compiler-vapor): ref for `v-for`

### PR DESCRIPTION
**Pull request by Jevon617**
Before:

Fixed:

### CODE REVIEW

File: packages/compiler-vapor/src/transforms/transformRef.ts [hunk 1]
```diff
@@ -10,7 +11,10 @@ import { findProp } from '../utils'
 import { EMPTY_EXPRESSION } from './utils'
 
 export const transformRef: NodeTransform = (node, context) => {
-  if (node.type !== NodeTypes.ELEMENT) return
+  // If the current node has conditional directives,
+  // it will result in duplicate registration of SET_REF operations.
+  if (node.type !== NodeTypes.ELEMENT || hasConditionalDirective(node)) return
+
```
Comment (LittleSound) [inline packages/compiler-vapor/src/transforms/transformRef.ts:17]: The same problem also applies to `v-for`, and the reason for the issue is more likely due to an incorrect order of Transform Preset. After I adjusted the order, the problem disappeared.

image

**Review by LittleSound — CHANGES_REQUESTED** [review]

File: packages/compiler-vapor/src/transforms/transformRef.ts (hunk 1 above)
Comment (Jevon617) [reply packages/compiler-vapor/src/transforms/transformRef.ts:17]: Oh, that's correct, but some of the behaviors of v for+ref don't seem to align with vue core. Perhaps another PR needs to
 
align these behaviors.

File: packages/runtime-vapor/src/dom/templateRef.ts [hunk 2]
```diff
@@ -31,7 +38,29 @@ export function setRef(el: Element, ref: NodeRef) {
...
-        if (_isString) {
+        if (ref_for) {
+          const existing = _isString
+            ? hasOwn(setupState, ref)
+              ? setupState[ref]
+              : refs[ref]
+            : ref.value
+          if (isUnmounted) {
+            isArray(existing) && remove(existing, value)
+          } else {
```
Comment (LittleSound) [inline packages/runtime-vapor/src/dom/templateRef.ts:49]: `doSet` only runs once at create, so this piece of code cannot remove the item that has already been removed in `v-for`. (this situation also exists in `v-if`)

See this [Playground](https://deploy-preview-167--vapor-repl.netlify.app/#__VAPOR__eNp9U02PmzAQ/Ssj9wBIFA7tKYJU7SqH7aGt2qqXugcCw8fG2JZt2Kwi/nvHkLC0Wi1SFPvNm+c34/GFfdQ6GQdkO5bZ0nTagUU36D2XXa+VcXABg3UMDbq7wRiU7l5aV8gSY3gsXNke6hpLBxPURvUQkFbAJZelIppPtZD7vzC6Yb1trlDQohAqoAiXG60wjCDfg2crgYlQTRh4oV0Qz4LJWIgBPyQCZePaLRaRVJYudVAFtHHYa1E4pB1A1kk9OBjf9qpCkXNGVjiDdAn6moG+WW7RhjzPITvuL3MT7O1ImKYsPS5Z6ZK2Zv/OjnRArQzJh2N8ikDVvmQ6Z3fCJ0JPtCQ1WnlN2nj5kTRjL/rnWTVLN+5ZzJylltRdkzxYJenCLp7KWal63Qk0X7XrqGWc7WCO+FhB/X38PGPODBjf8LLF8vQC/mDPHuPsm0GLZkTO1pgrDA3BEj78+IJnWq9B6uggiP1K8DvSdQ7e40L7NMiKbG94s9v7eew62fy0h7NDaW9FeaOeOc18zmjS7l4p/dnuu+T9nMflRF3sZIXnpHW92My8e9Lo5+FaxfUu/be+gtIg3cSvgrb0Zrbzno4epKn/L8fTFlKSXp/ZhvSvXki/KOnVIF0YvCm0DqJ1kLOqG6GryB/hZC5LCdiz6S8BOTlY)

Comment (Jevon617) [reply packages/runtime-vapor/src/dom/templateRef.ts]: It may be feasible to execute refs remove in the callback of `onScopeDispose`.

**Review by LittleSound — APPROVED** [review]

**Review by sxzz — APPROVED** [review]

**Pull request merged**